### ISSUE TITLE
Suggestion: Move all grammar-related metadata to `grammars.yml`

### ISSUE THREAD

**Issue by Alhadis** [github/linguist#4990]
Currently, metadata pertaining to grammars and case-by-case exceptions are handled in four different places:

* [`.gitmodules`](https://github.com/github/linguist/blob/b7f579597050aaf7ac19c6c6ec504253a577ff13/.gitmodules) specifies URLs.
* [`grammars.yml`](https://github.com/github/linguist/blob/b7f579597050aaf7ac19c6c6ec504253a577ff13/grammars.yml) specifies scopes.
* [`tools/grammars/compiler/data.go`](https://github.com/github/linguist/blob/16582708fe8a35883862143fbdf5eb1ec843d6f1/tools/grammars/compiler/data.go) specifies scope-maps and case-by-case overrides
* [`vendor/licenses/config.yml`](https://github.com/github/linguist/blob/96ad1185828f44bb9b774328a584551ee57ed264/vendor/licenses/config.yml#L16-L22) identifies (for Licensee) which grammars had their licenses manually-reviewed

Moreover, the difficulties [related](https://github.com/github/linguist/issues/4974#issuecomment-684577601) by @pastra98 made me realise there's more we could be doing with regards to locating grammar and license files. Specifically, we should be able to provide a manual path if need be — the [currently hardcoded](https://github.com/github/linguist/blob/ec800d3887d8d635b7a1d9053076bee35431b757/tools/grammars/compiler/loader.go#L106-L129) search locations can remain the default for grammars without a `src:` field defined, or whatever.

<details><summary>Here's how it *might* look.</summary>

~~~yml
# Each entry correlates to a directory in "vendor/grammars/#{key}"
--
abl-tmlanguage:
    license: MIT
    source: chriscamicas/abl-tmlanguage
    scopes:
        - source-abl
    
    # Location of files inside submodule repository. Usually
    # calculated automatically, though maybe it makes sense
    # to always provide this list, as opposed to only those
    # grammars with non-standard file locations?
    files:
        grammar: abl.tmLanguage.json
        license: LICENSE

actionscript3-tmbundle:
    license: MIT
    source: simongregory/actionscript3-tmbundle
    scopes:
        - source.actionscript.3
        - text.html.asdoc
        - text.xml.flex-config

c.tmbundle:
    license: MIT
    source: textmate/c.tmbundle
    scopes:
        - source.c
        - source.c++
        - source.c.platform
    aliases:
        source.c++: source.cpp

hy.tmLanguage:
    license: MIT
    source: Slowki/hy.tmLanguage
    scopes:
        - source.hy
    paths:
        - hy.json
        - LICENSE.md

language-roff:
    license: ISC
    source: Alhadis/language-roff
    scopes:
        - hidden.manref
        - source.ditroff
        - source.ditroff.desc
        - source.gremlin
        - source.ideal
        - source.pic
        - text.roff
        - text.runoff

sublimesystemverilog:
    license: MIT
    source: https://bitbucket.org/Clams/sublimesystemverilog/get/default.tar.gz
    scopes:
        - source.systemverilog
        - source.ucfconstraints

Genshi.tmbundle:
    license: MIT
    source: https://svn.edgewall.org/repos/genshi/contrib/textmate/Genshi.tmbundle/Syntaxes/Markup%20Template%20%28XML%29.tmLanguage
    scopes:
        - text.xml.genshi
~~~

</details>

Thoughts?

**Comment by pastra98** [comment]
That would be pretty cool, actually! Maybe as a flag with the `add-grammar` script to specify the path?

**Comment by Alhadis** [comment]
No need. Most of the time, the files are located somewhere predictable. Atom actually *forces* you to place them in the `grammars` directory, and limits you to JSON or CSON (a cleaner alternative to JSON) (it also imposes a bunch of other myopic, insane restrictions which I won't go into here).

**Comment by pastra98** [comment]
Ah okay... So it's not that many repos that are structured differently. But I could add a `source: ` field to the `grammars.yml` manually, if need be?

**Comment by Alhadis** [comment]
> But I could add a `source:` field to the `grammars.yml` manually, if need be?

This is more of an RFC to discuss a *potential* enhancement to Linguist. The changes involved are many, and they touch lots of different components that behave similarly, but operate independently.

In other words, there's nothing you need to worry about WRT to your PR. 😉

**Comment by pastra98** [comment]
I get that, I was trying to understand how you envision this to work from a users perspective haha

**Comment by Alhadis** [comment]
The `add-grammar` script would support an optional `-f`/`--file` switch to specify the location of a grammar file, but that's about it. Usage would be something like `--file hy.json` or `--file ./hy.json` or even

	--file https://github.com/Slowki/hy.tmLanguage/blob/master/hy.json

since users will try all sorts of things, and it's better for a program to be maximally permissive about what input it accepts, rather than barking at users _"to provide a path relative to the upstream repository's root directory"_.

@lildude I don't think I've overdone that script enough, should I add a man page? 😁

**Comment by lildude** [comment]
This sounds like an interesting idea, though is likely to be a lot of work. Is all that work really worth the effort for the few corner cases not correctly handled now?

Without thinking too hard about it, a few quick points come to mind:

- we won't be able to remove the dependency on submodules unless you implement a script to do the downloading and updating... which would be reinventing what git already does which could be more brittle and harder to debug as it would require more specific knowledge.
- I'm not sure you'll be able to get rid of `vendor/licenses/config.yml` whilst retaining the required Licensed/Licensee integration, without implementing some more custom code, which I'd prefer we didn't do as we've only just got back to using licensed as it is intended to be used.

So if anything, this change would really be consolidating some of the current `grammars.yml` and `tools/grammars/compiler/data.go` info.

Feel free to start a PoC PR and we can track and discuss things as it progresses.

**Comment by johnmays** [comment]
Why are `grammars.yml` and `languages.yml` in two completely different spots?  Wouldn't it make sense to move `grammars.yml` out of root and into `/lib/linguist`?

**Comment by lildude** [comment]
@johnmays Not really. `grammars.yml` isn’t actually used directly by Linguist. It’s used to track the external grammars which are used by the syntax highlighting engine which is a completely independent application. The same applies to the `vendor/licenses/config.yml` - it’s not directly used by linguist and in that case is used to configure licensed, the external library we use for license management.

**Comment by johnmays** [comment]
I see. That makes sense.
